hypergan/losses/base_loss.py

```diff
@@ -90,22 +96,42 @@ def half(tens, second):
 
         return ops.squash(ops.concat([f1, f2]))
 
+    def gradient_locally_stable(self):
+        discriminator = self.discriminator or self.gan.discriminator
+        config = self.config
+        generator = self.generator or self.gan.generator
+        g_sample = self.gan.uniform_sample
+        gradients = tf.gradients(discriminator.sample, [g_sample])[0]
+        penalty = tf.sqrt(tf.reduce_sum(tf.square(gradients), axis=1))
+        penalty = tf.reduce_mean(tf.square(penalty))
+        return float(config.gradient_locally_stable) * penalty
+
     def gradient_penalty(self):
         config = self.config
         gan = self.gan
+        ops = self.gan.ops
         gradient_penalty = config.gradient_penalty
-        if has_attr(gan.inputs, 'gradient_penalty_label'):
+        x = gan.inputs.x
+        if config.gradient_penalty_label is not None:
+            x = config.gradient_penalty_label
+        if hasattr(gan.inputs, 'gradient_penalty_label'):
             x = gan.inputs.gradient_penalty_label
-        else:
-            x = gan.inputs.x
+
         generator = self.generator or gan.generator
         g = generator.sample
         discriminator = self.discriminator or gan.discriminator
         shape = [1 for t in g.get_shape()]
         shape[0] = gan.batch_size()
         uniform_noise = tf.random_uniform(shape=shape,minval=0.,maxval=1.)
         print("[gradient penalty] applying x:", x, "g:", g, "noise:", uniform_noise)
-        interpolates = x + uniform_noise * (g - x)
+        if config.gradient_penalty_type == 'dragan':
+            axes = [0, 1, 2, 3]
+            if len(ops.shape(x)) == 2:
+                axes = [0, 1]
+            mean, variance = tf.nn.moments(x, axes=axes)
+            interpolates = x + uniform_noise * 0.5 * variance * tf.random_uniform(shape=ops.shape(x), minval=0.,maxval=1.)
```

Comment:
Maybe we should treat the second random uniform as an encoder that defaults to the paper.